src/cargo/core/compiler/job_queue.rs

```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                        info!("end: {:?}", id);
+                        self.finished += 1;
+                        if let Some(rustc_tokens) = self.rustc_tokens.remove(&id) {
+                            // This puts back the tokens that this rustc
+                            // acquired into our primary token list.
+                            //
+                            // FIXME: this represents a rustc bug: it did not
+                            // release all of its thread tokens but finished
+                            // completely.
+                            self.tokens.extend(rustc_tokens);
+                        }
+                        while let Some(pos) =
+                            self.to_send_clients.iter().position(|(i, _)| *i == id)
+                        {
+                            // drain all the pending clients
+                            self.to_send_clients.remove(pos);
+                        }
+                        self.active.remove(&id).unwrap()
+                    }
+                    // ... otherwise if it hasn't finished we leave it
+                    // in there as we'll get another `Finish` later on.
+                    Artifact::Metadata => {
+                        info!("end (meta): {:?}", id);
+                        self.active[&id]
+                    }
+                };
+                info!("end ({:?}): {:?}", unit, result);
+                match result {
+                    Ok(()) => self.finish(id, &unit, artifact, cx)?,
+                    Err(e) => {
+                        let msg = "The following warnings were emitted during compilation:";
+                        self.emit_warnings(Some(msg), &unit, cx)?;
+
+                        if !self.active.is_empty() {
+                            handle_error(&e, &mut *cx.bcx.config.shell());
+                            cx.bcx.config.shell().warn(
+                                "build failed, waiting for other \
+                                 jobs to finish...",
+                            )?;
+                            return Ok(Some(anyhow::format_err!("build failed")));
+                        } else {
+                            return Ok(Some(e));
+                        }
+                    }
+                }
+            }
+            Message::Token(acquired_token) => {
+                let token = acquired_token.chain_err(|| "failed to acquire jobserver token")?;
+                self.tokens.push(token);
+            }
+            Message::NeedsToken(id, client) => {
+                log::info!("queue token request");
+                jobserver_helper.request_token();
+                self.to_send_clients.push((id, client));
+            }
+            Message::ReleaseToken(id) => {
+                // Note that this pops off potentially a completely
+                // different token, but all tokens of the same job are
+                // conceptually the same so that's fine.
```

Comment:
Could this comment be extended a bit as well? I think this all works because `self.tokens` is largely a "pool" of tokens which we greedily allocate requests from, but we also truncate it before we block on all iterations of the loop. That way we're basically transferring ownership of the token through the `self.tokens` list